internal/website/gatherexamples/gatherexamples.go

```diff
@@ -0,0 +1,301 @@
+	"flag"
+	"fmt"
+	"go/ast"
+	"go/printer"
+	"go/types"
+	"os"
+	"sort"
+	"strings"
+
+	"golang.org/x/tools/go/packages"
+)
+
+func main() {
+	flag.Usage = func() {
+		out := flag.CommandLine.Output()
+		fmt.Fprintln(out, "usage: gatherexamples [options] DIR [...]")
+		fmt.Fprintln(out)
+		fmt.Fprintln(out, "Options:")
+		flag.PrintDefaults()
+	}
+	outputPath := flag.String("o", "", "path to output file (default to stdout)")
+	pattern := flag.String("pattern", "./...", "Go package pattern to use at each directory argument")
+	flag.Parse()
+	if flag.NArg() == 0 {
+		flag.Usage()
+		os.Exit(2) // matches with flag package
+	}
+
+	// Load packages in each module named on the command line and find
+	// all examples.
+	allExamples := make(map[string]string)
+	for _, dir := range flag.Args() {
+		cfg := &packages.Config{
+			Mode:  gatherLoadMode,
+			Dir:   dir,
+			Tests: true,
+		}
+		pkgs, err := packages.Load(cfg, *pattern)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: load %s: %v\n", dir, err)
+			os.Exit(1)
+		}
+		for exampleName, source := range gather(pkgs) {
+			allExamples[exampleName] = source
+		}
+	}
+
+	// Write all examples as a JSON object.
+	data, err := json.MarshalIndent(allExamples, "", "\t")
+	if err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: generate JSON: %v\n", err)
+		os.Exit(1)
+	}
+	data = append(data, '\n')
+
+	output := os.Stdout
+	if *outputPath != "" && *outputPath != "-" {
+		var err error
+		output, err = os.Create(*outputPath)
+		if err != nil {
+			fmt.Fprintf(os.Stderr, "gatherexamples: %v\n", err)
+			os.Exit(1)
+		}
+		defer func() {
+			if err := output.Close(); err != nil {
+				fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+				os.Exit(1)
+			}
+		}()
+	}
+	if _, err := output.Write(data); err != nil {
+		fmt.Fprintf(os.Stderr, "gatherexamples: write output: %v\n", err)
+		os.Exit(1)
+	}
+}
+
+const gatherLoadMode packages.LoadMode = packages.NeedName |
+	packages.NeedFiles |
+	packages.NeedTypes |
+	packages.NeedSyntax |
+	packages.NeedTypesInfo
+
+// signifierCommentPrefix is the start of the comment used to signify whether
+// the example should be included in the output.
+const signifierCommentPrefix = "// This example is used in"
+
+// gather extracts the code from the example functions in the given packages
+// and returns a map like the one described in the package documentation.
+func gather(pkgs []*packages.Package) map[string]string {
+	examples := make(map[string]string)
+	for _, pkg := range pkgs {
+		for _, file := range pkg.Syntax {
+			for _, decl := range file.Decls {
+				// Determine whether this declaration is an example function.
+				fn, ok := decl.(*ast.FuncDecl)
+				if !ok {
+					continue
+				}
+				if !ok || !strings.HasPrefix(fn.Name.Name, "Example") || len(fn.Type.Params.List) > 0 || len(fn.Type.Params.List) > 0 {
+					continue
+				}
+
+				// Gather sorted list of imported packages.
+				usedPackages := make(map[string]string)
```

Comment:
The comment says "sorted list" but it's a map.

Separately, if it comes as a sorted list couldn't we save it in order and save a bunch of the work that `formatImports` is doing?

Reply:
Comment was stale: it does not come in a sorted list, but an earlier iteration of my code would do the sorting in `formatImports` pretty close to the comment. Since then, the code grew.